cctp_primitives/src/proving_system/init.rs

```diff
@@ -18,104 +13,204 @@ use std::sync::{RwLock, RwLockReadGuard};
+/// If `G2_UNIVERSAL_PARAMETERS` has been initialized, return `CommitterKeyG2`, otherwise return
+/// Error.
+/// If `supported_degree.is_some()`, then `CommitterKeyG2` is trimmed to the specified size.
+pub fn get_g2_committer_key(
+    supported_degree: Option<usize>,
+) -> Result<CommitterKeyG2, ProvingSystemError> {
+    let pp_g2_guard = G2_UNIVERSAL_PARAMS.read().map_err(|_| {
+        ProvingSystemError::Other("Failed to acquire lock for G2_UNIVERSAL_PARAMS".to_owned())
     })?;
-    if ck_g2_guard.is_some() {
-        Ok(ck_g2_guard)
+
+    if pp_g2_guard.is_some() {
+        let supported_degree =
+            supported_degree.unwrap_or_else(|| pp_g2_guard.as_ref().unwrap().max_degree());
+        let (ck, _) =
+            InnerProductArgPC::<_, Digest>::trim(&pp_g2_guard.as_ref().unwrap(), supported_degree)
+                .map_err(|err| ProvingSystemError::Other(err.to_string()))?;
+        Ok(ck)
     } else {
         Err(ProvingSystemError::CommitterKeyNotInitialized)
     }
 }
 
-fn load_generators<G: AffineCurve>(
+fn load_universal_params<G: AffineCurve>(
     max_degree: usize,
-    supported_degree: usize,
-) -> Result<CommitterKey<G>, SerializationError> {
+) -> Result<UniversalParams<G>, SerializationError> {
     let pp = InnerProductArgPC::<G, Digest>::setup(max_degree)
         .map_err(|_| SerializationError::InvalidData)?;
-    let (ck, _) = InnerProductArgPC::<G, Digest>::trim(&pp, supported_degree)
-        .map_err(|_| SerializationError::InvalidData)?;
 
-    // Return the read/generated committer key
-    Ok(ck)
+    // Return the read/generated universal parameters
+    Ok(pp)
+}
+
+#[cfg(test)]
+pub(crate) fn max_pow_for_testing() -> usize {
+    return 10;
+}
+
+#[cfg(test)]
+/// In normal usage, `load_g{1,2}_universal_params()` should be called exactly once during program
+/// execution, because further calls to these functions generate an error.
+/// This pattern of usage however is not ideal in unit tests, because:
+/// - the unit test containing the call to `load_g{1,2}_universal_params()` would need to be run
+///   before the other unit tests using the committer key. This would generate an unwanted ordering
+///   constraint in unit tests execution and interfere with parallel test execution.
+/// - the unit tests using the committer key would pass when run in batch, but fail when run
+///   individually.
+/// To overcome this difficulty, the functions `set_g{1,2}_universal_params()` are provided. They
+/// are simple wrappers around `load_g{1,2}_universal_params()`, fixing a uniform value for
+/// `max_degree` and catching (and discarding) the error generated by additional calls to
+/// `load_g{1,2}_universal_params()`.
+///
+/// Every unit test which needs to access committer keys should call `set_g{1,2}_universal_params()`
+/// in advance. This solves the aforementioned problems, by ensuring that the committer key is
+/// initialized before being accessed.
```

Comment:
Isn't it simpler to initialize the keys with the same max degree in all tests (to ensure no inconsistencies across tests) and simply ignore the error if the keys have already been initialized by another test ?